packages/feature-detector/src/utils/functions.ts

```diff
@@ -0,0 +1,25 @@
+import { Node, SyntaxKind, Type } from 'ts-morph'
+
+const { ExportAssignment, Identifier, FunctionDeclaration, ArrowFunction } =
+  SyntaxKind
+
+/**
+ * Exteracts the return type of a function, if any.
+ * Supports:
+ * - function declaration `function foo() {}`
+ * - function expression `const foo = function() {}`
+ * - arrow functions `const foo = () => {}`
+ * - module exports
+ */
+export function getReturnType(node?: Node): Type | undefined {
+  switch (node?.getKind()) {
+    case ExportAssignment:
+      return getReturnType(node.asKind(ExportAssignment)?.getExpression())
```

Comment:
question: May I assume that getReturnType won't be in an infinite loop? Is the following code properly rejected?

```typescript
type A = A;
function circular(): A {}
```

Reply:
I added two test cases. For [invalid cyclic types](https://www.typescriptlang.org/play?#code/C4TwDgpgBAglC8sDcVVvR1B6LUCWAdgG4CGANngCYBcUAxngE50CuZJjUokUlEAZoTzA8AewIAoCfxYE6I8VH6jRACgCUtOAG8AvkA), `getReturnType()` would return `any`.

For [valid cyclic reference](https://www.typescriptlang.org/play?#code/MYewdgzgLgBApgDwA4gE5TgExgXhgCgEpcA+GAbxipgHoaYA3AQwBsBLTALhmAE9h2wGKjgAzOCLDA4EAFBVQkWAlwUAvvJgIAdGESw8CTSKgBXVGC279AbllqgA), it depends on the language: `any` for TS, and `{ next: any }` for JS (interesting!).